### ISSUE TITLE
Add a property to read container average bitrate

### ISSUE THREAD

**Issue by agyild** [mpv-player/mpv#8995]
Basically what I wish to accomplish is to calculate `bits(pixel*frame)` value of a video file and pick a profile according to the theoretical quality of the video based on its average bitrate, resolution and frame rate reported by the container. However I was unable to find a way to read the container's video average bitrate via properties. The closest thing I could find was `video-bitrate` which is not an average value and unfit for profile selection (because most videos start with a fade-in or something similar which requires less bitrate at the beginning). There was also `video-params/average-bpp` but I think it is for color depth since it displayed `24` for all the videos that I have tested. I would appreciate it if you could offer me a workaround or consider adding it as a feature.

**Comment by TheAMM** [comment]
Divide the filesize by the duration. I'm not sure where you're doing this, but you should be capable of a little math. If you have matroska files muxed by mkvtoolnix, you may have luck in getting the video stream's `NUMBER_OF_BYTES` metadata value, if that's important.

**Comment by Dudemanguy** [comment]
`file-size` and `duration` exist as noted so that should be sufficient.

**Comment by agyild** [comment]
> `file-size` and `duration` exist as noted so that should be sufficient.

It should be noted that this is only a workaround and will give inaccurate results when the container's non-video streams take up more space. For this reason I'd you ask to reconsider. There is actually a very good use case for this metric and it should be relatively easy to implement, just reading a few metadata bytes from the file. I am not familiar with mpv codebase to implement it myself but If it's possible to get raw file access with lua I can also do it with a script.

**Comment by Dudemanguy** [comment]
My apologies, I thought you wanted the bitrate of the whole file (aka `file-size`/`duration) and not individual streams. I guess in theory it would be a part of the `track-list` property. There is currently `track-list/N/demux-bitrate` although that is audio only (according to the documentation anyway) and wouldn't cover what you're asking (also it seems to be unreliable from some brief testing). I don't know the feasibility of making that value actually useful but it's a valid feature request.

**Comment by agyild** [comment]
Yes, I meant the individual video streams. In my case for profile selection I'd just use the default video stream's average bitrate to calculate the bits per pixel value. Here is an excerpt from a MediaInfo output:

```
Video
ID                          : 1
Format                      : HEVC
Format/Info                 : High Efficiency Video Coding
Format profile              : Main 10@L4@Main
Codec ID                    : V_MPEGH/ISO/HEVC
Duration                    : 2 h 41 min
Bit rate                    : 1 501 kb/s
Width                       : 1 920 pixels
Height                      : 800 pixels
Display aspect ratio        : 2.40:1
Frame rate mode             : Constant
Frame rate                  : 23.976 (24000/1001) FPS
Color space                 : YUV
Chroma subsampling          : 4:2:0
Bit depth                   : 10 bits
Bits/(Pixel*Frame)          : 0.041
```

`Bits/(Pixel*Frame)` value is calculated with the following formula: `average_bitrate / (width * height * framerate) * 1000` and then rounded down or up. Every encoder and encoding mode usually has different characteristics but it should give one an approximate idea about the possible quality of the video stream.

**Comment by TheAMM** [comment]
> just reading a few metadata bytes from the file

As said, matroska *may* contain this information, but I'm not sure *all* containers do. mpv would have to demux the entire file and sum up the packet sizes per stream, which IMO is unacceptable. Going by the first few (mega)bytes we've seen is also not proper, as noted. AFAIK mediainfo parses the entire file, or relies on format-specific metadata which you may not be able to rely on.

You can write a Lua script to call ffprobe to go through the entire file and dump packet sizes, ala `ffprobe -of compact=p=0 -show_packets video.mkv`. You can limit yourself only to the pertinent fields with something like `packet=stream_index,size`.
If you'll write this script, I recommend using Lua's io.popen to access the stdout and stream the lines instead of collecting megabytes of the stuff with the subprocess command. IIRC it's also possible to limit the amount of data (by bytes or seconds) ffprobe will read, if you figure you don't need to scan through 15GiBs of movie to decide on a bitrate profile.

**Comment by agyild** [comment]
I think it is an acceptable compromise to return the container reported avg. bitrate or otherwise return *(unspecified)* if the majority of files contain it as metadata (Not sure but don't the majority of mkv/mp4 containers have it?). And in other cases one might use ffprobe to get the information as you have shown.

Manual calculation by mpv should be doable in my opinion. What about getting an approximate bitrate by probing the video stream at multiple random points and dividing the total bitrate by the number of probes? (so you don't have to decode the entire stream) It could be made more accurate by making sure that the probed frame has sufficient entropy in it (e.g. it is not a single colored frame). And finally you can calculate the value lazily for better performance. This approach although might require some work to implement, is more bullet-proof in my opinion since it doesn't depend on any optional container metadata.

And for the sake of completion, for livestreams you can just report back the average bitrate of the initially buffered stream.

What are your thoughts? In the meanwhile I'll try to do something by scripting.

**Comment by TheAMM** [comment]
> if the majority of files contain it as metadata

Eh, depends. Again, mkvtoolnix-muxed matroska files are likely to have `NUMBER_OF_BYTES` tags for the streams, though ffmpeg nor mpv check those. MP4/MOV/etc have... various atoms, possibly, from which the average bitrate can be deduced, and ffmpeg does do this? but mpv doesn't copy the `codec->bit_rate` on video tracks.
A quick test adding `sh->codec->bitrate = codec->bit_rate;` in `handle_new_stream()` for the lavf demuxer did at least show some values for video tracks, but one would have to check all demuxers for what can be done. And since it's trivial to copy the field, itshould be considered *why* it's not being copied already...

> Manual calculation by mpv should be doable in my opinion
> probing the video stream at multiple random points

Technically doable, of course. Reasonable given the complexity, potential breakage and niche usecase? No, no! Making *mpv* do random seeks to probe the file for a bitrate sound horrible!
This really is something to implement as a script, if you want foolproof stream bitrates. You can even do your seek optimization with ffprobe.

**Comment by Hrxn** [comment]
> Bits/(Pixel*Frame)          : 0.041
> [..]
> `Bits/(Pixel*Frame)` value is calculated with the following formula: `average_bitrate / (width * height * framerate) * 1000` and then rounded down or up. Every encoder and encoding mode usually has different characteristics but it should give one an approximate idea about the possible quality of the video stream.

I mean, it depends. It *can* be a useful heuristic, but only if you understand what it actually means. In other words, if you already know the underlying concepts behind it. Because this value alone does not tell you enough, Bits/(Pixel*Frame) can be pretty high for some old or even ancient video codec, but the quality might still not be different (or even worse) to the results a modern video compression technology provides, by pretty big margins, actually. Also, you need to be aware of the image dimensions, because you can only directly compare files with (roughly) the same values in terms of dimensions.
